cms/src/main/java/uk/nhs/digital/externalstorage/resource/ResourceUploadPlugin.java

```diff
@@ -0,0 +1,114 @@
+import uk.nhs.digital.externalstorage.s3.S3Connector;
+import uk.nhs.digital.externalstorage.s3.S3ObjectDescriber;
+
+import java.io.IOException;
+import java.util.Calendar;
+
+import javax.jcr.Node;
+import javax.jcr.RepositoryException;
+
+
+/**
+ * Plugin for uploading resources into the JCR repository.
+ * This plugin can be configured with specific types, so not all file types are allowed to be uploaded.
+ */
+public class ResourceUploadPlugin extends RenderPlugin {
+
+    static final Logger log = LoggerFactory.getLogger(ResourceUploadPlugin.class);
+    public static final String DEFAULT_ASSET_VALIDATION_SERVICE_ID = "service.gallery.asset.validation";
+
+    private final IEditor.Mode mode;
+
+    public ResourceUploadPlugin(IPluginContext context, IPluginConfig config) {
+        super(context, config);
+        mode = IEditor.Mode.fromString(config.getString("mode"), IEditor.Mode.EDIT);
+        add(createFileUploadPanel());
+        add(new EventStoppingBehavior("onclick"));
+    }
+
+    private FileUploadPanel createFileUploadPanel() {
+        final FileUploadPanel panel = new FileUploadPanel("fileUpload", getPluginConfig(), getValidationService()) {
+            @Override
+            public void onFileUpload(final FileUpload fileUpload) throws FileUploadViolationException {
+                handleUpload(fileUpload);
+            }
+        };
+        panel.setVisible(mode == IEditor.Mode.EDIT);
+        return panel;
+    }
+
+    private FileUploadValidationService getValidationService() {
+        return DefaultUploadValidationService.getValidationService(getPluginContext(), getPluginConfig(),
+            DEFAULT_ASSET_VALIDATION_SERVICE_ID);
+    }
+
+    /**
+     * Handles the file upload from the form.
+     *
+     * @param upload the {@link FileUpload} containing the upload information
+     */
+    private void handleUpload(FileUpload upload) throws FileUploadViolationException {
+        S3Connector s3Connector = HippoServiceRegistry.getService(S3Connector.class);
+        S3ObjectDescriber describer = new S3ObjectDescriber(s3Connector.getBucketName());
+        String fileName = upload.getClientFileName();
+        String mimeType = upload.getContentType();
+
+        JcrNodeModel nodeModel = (JcrNodeModel) this.getDefaultModel();
+        Node node = nodeModel.getNode();
+        try {
+            String s3ObjectKey = describer.generateObjectKey(fileName);
+
+            // InputStream inputStream
+            s3Connector.uploadFile(upload.getInputStream(), s3ObjectKey);
+
+            setResourceProperties(node, mimeType, fileName);
+            decorateNode(node, s3ObjectKey, describer.getUrl(s3ObjectKey), s3Connector.getFileSize(s3ObjectKey));
+
+            if (node.hasProperty(HIPPO_TEXT)) {
+                node.getProperty(HIPPO_TEXT).remove();
+            }
+        } catch (RepositoryException | IOException ex) {
+            if (log.isDebugEnabled()) {
```

Comment:
It feels like it's okay to always log an error, even when re-throwing the exception - unless the exception is handled and logged sufficiently well by Wicket (that is, preserving and logging the entire stack trace associated with the original exception).

Either way, the `else` clause shouldn't be required - the first of the `log.error` statements will print out the cause exception's message in the stack trace.

Reply:
copy & paste code for hippo. Updated now :)